### PR TITLE
[Networking] Implements Pluggable Conduit Factory

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR refactors the networking layer to decouple the creation of `Conduit`s into a stateful `ConduitFactory` interface, which provides a _pluggable_ `Conduit` creation functionality for the networking layer. This is a prerequisite decomposition to enable testing our nodes under Byzantine behavior through the _Programmable Conduit (PC)_ framework (see the attached issue for more details). 

In a nutshell, a `Conduit` is the interface that enables an engine push a message to a subset of the remote nodes through the networking layer.

### CODE REVIEW

**Review by vishalchangrani — APPROVED** [review]

File: network/conduit.go [hunk 1]
```diff
@@ -3,57 +3,23 @@
...
-// Nodes subscribed to the same channel can disseminate epidemic messages among
-// each other, i.e.. multicast and publish.
-type Channel string
-type ChannelList []Channel
+// ConduitFactory is an interface type that is utilized by the Network to create conduits for the channels.
+type ConduitFactory interface {
+	// RegisterAdapter sets the Adapter component of the factory.
+	// The Adapter is a wrapper around the Network layer that only exposes the set of methods
+	// that are needed by a conduit.
+	RegisterAdapter(Adapter) error
```
Comment (gomisha) [inline network/conduit.go:18]: so we'll be able to pass different `Adapter`s to the `ConduitFactory` depending on whether it's the real or the testing `Adapter`?

A full diagram showing how the `Adapter`, `ConduitFactory`, `Conduit`, `Channel` and the network fit together would be helpful

File: network/channel.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package network
```
[the remark this answers is not in the document]
Comment (gomisha) [reply network/channel.go:1]: Can you comment on why the channel logic was moved into a separate file? Is it to allow test channels to be created to simulate BFT attack scenarios?

File: network/conduit.go [hunk 3]
```diff
@@ -3,57 +3,23 @@
...
-func (cl ChannelList) Contains(channel Channel) bool {
-	for _, c := range cl {
-		if c == channel {
-			return true
-		}
-	}
-	return false
+	// NewConduit creates a conduit on the specified channel.
+	// Prior to creating any conduit, the factory requires an Adapter to be registered with it.
+	NewConduit(context.Context, Channel) (Conduit, error)
```
Comment (gomisha) [inline network/conduit.go:22]: can you clarify why we're specifying a `Channel` for the `Conduit`? Is it to allow specifying different channels depending on if it's a normal or a test channel?

**Review by gomisha — APPROVED** [review]

File: network/conduit.go (hunk 1 above)
Comment (yhassanzadeh13) [reply network/conduit.go:18]: We resolved this in our meeting.

File: network/p2p/conduit/conduit.go [hunk 4]
```diff
@@ -8,32 +8,55 @@ import (
...
+}
+
+func NewDefaultConduitFactory() *DefaultConduitFactory {
+	return &DefaultConduitFactory{}
+}
+
+// RegisterAdapter sets the Adapter component of the factory.
+// The Adapter is a wrapper around the Network layer that only exposes the set of methods
+// that are needed by a conduit.
+func (d *DefaultConduitFactory) RegisterAdapter(adapter network.Adapter) error {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/conduit/conduit.go:25]: We resolved this in our meeting.

File: network/conduit.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/conduit.go:22]: We resolved this in our meeting.

**Pull request merged**